### PR TITLE
Add TLS support to auto-instrumentation

### PR DESCRIPTION
**Pull request by pavolloffay**
**Description:** <Describe what has changed.>

**Link to tracking Issue(s):** <Issue number if applicable>

- Resolves: #2997

Todos:
- [x] e2e tests
- [x] more unit tests
- [x] follow up issue with proposal for cert-manager - https://github.com/open-telemetry/opentelemetry-operator/issues/3342 

**Testing:** <Describe what testing was performed and which tests were added.>

On OpenShifit
```
kubectl apply -f - <<EOF
apiVersion: opentelemetry.io/v1beta1
kind: OpenTelemetryCollector
metadata:
  name: simplest
  annotations:
    service.beta.openshift.io/serving-cert-secret-name: otelcol
spec:
  volumeMounts:
  - name: otelcol-certs
    mountPath: /otelcol-certs
  volumes:
  - name: otelcol-certs
    secret: 
      secretName: otelcol
    
  config:
    receivers:
      otlp:
        protocols:
          grpc:
            endpoint: 0.0.0.0:4317
            tls:
              cert_file: /otelcol-certs/tls.crt
              key_file: /otelcol-certs/tls.key
          http:
            endpoint: 0.0.0.0:4318
            tls:
              cert_file: /otelcol-certs/tls.crt
              key_file: /otelcol-certs/tls.key
    processors:
    exporters:
      debug: {}

    service:
      pipelines:
        traces:
          receivers: [otlp]
          processors: []
          exporters: [debug]
EOF


kubectl apply -f - <<EOF
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: my-instrumentation
spec:
  exporter:
    endpoint: https://simplest-collector.tracing-system.svc.cluster.local:4317
    tls:
      ca: /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
  propagators:
    - tracecontext
    - baggage
    - b3
  sampler:
    type: parentbased_traceidratio
    argument: "1"
EOF

kubectl apply -f https://raw.githubusercontent.com/pavolloffay/kubecon-eu-2024-opentelemetry-kubernetes-tracing-tutorial/main/app/k8s.yaml
kubectl patch deployment frontend-deployment -n tutorial-application -p '{"spec": {"template":{"metadata":{"annotations":{"instrumentation.opentelemetry.io/inject-sdk":"tracing-system/my-instrumentation"}}}} }'
kubectl patch deployment backend1-deployment -n tutorial-application -p '{"spec": {"template":{"metadata":{"annotations":{"instrumentation.opentelemetry.io/inject-python":"tracing-system/my-instrumentation"}}}} }'
kubectl patch deployment backend2-deployment -n tutorial-application -p '{"spec": {"template":{"metadata":{"annotations":{"instrumentation.opentelemetry.io/inject-java":"tracing-system/my-instrumentation"}}}} }'
k get pods -n tutorial-application -w 
kubectl port-forward service/frontend-service -n tutorial-application 4000:4000 
```

Another Example for Openshift Ca bundle and serving certs
```yaml
kubectl apply -f - <<EOF
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoMonolithic
metadata:
  name: tempo
spec:
  storage:
    traces:
      backend: memory 
      size: 1Gi 
  ingestion:
    otlp:
      grpc:
        enabled: true
        tls:
          enabled: true
      http:
        enabled: true
        tls:
          enabled: true
  jaegerui:
    enabled: true 
    route:
      enabled: true 
EOF


kubectl apply -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: otelcol-cabundle
  annotations:
    service.beta.openshift.io/inject-cabundle: "true"
EOF

kubectl apply -f - <<EOF
apiVersion: opentelemetry.io/v1beta1
kind: OpenTelemetryCollector
metadata:
  name: simplest
  annotations:
    service.beta.openshift.io/serving-cert-secret-name: otelcol
spec:
  volumeMounts:
  - name: otelcol-certs
    mountPath: /otelcol-certs
  - name: otelcol-ca
    mountPath: /otelcol-certs-ca
  volumes:
  - name: otelcol-certs
    secret: 
      secretName: otelcol
  - name: otelcol-ca
    configMap: 
      name: otelcol-cabundle
    
  config:
    receivers:
      otlp:
        protocols:
          grpc:
            endpoint: 0.0.0.0:4317
            tls:
              cert_file: /otelcol-certs/tls.crt
              key_file: /otelcol-certs/tls.key
          http:
            endpoint: 0.0.0.0:4318
            tls:
              cert_file: /otelcol-certs/tls.crt
              key_file: /otelcol-certs/tls.key
    processors:
    exporters:
      otlp:
        endpoint: https://tempo-tempo.tracing-system.svc.cluster.local:4317
        tls:
          ca_file: /otelcol-certs-ca/service-ca.crt
      debug: {}

    service:
      pipelines:
        traces:
          receivers: [otlp]
          processors: []
          exporters: [debug,otlp]
EOF



kubectl apply -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: otelcol-cabundle
  namespace: tutorial-application
  annotations:
    service.beta.openshift.io/inject-cabundle: "true"
EOF

kubectl apply -f - <<EOF
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: my-instrumentation
spec:
  exporter:
    endpoint: https://simplest-collector.tracing-system.svc.cluster.local:4317
    tls:
      configMapName: otelcol-cabundle
      ca: service-ca.crt
  propagators:
    - tracecontext
    - baggage
    - b3
  sampler:
    type: parentbased_traceidratio
    argument: "1"
  python:
    env:
      # Required if endpoint is set to 4317.
      # Python autoinstrumentation uses http/proto by default
      # so data must be sent to 4318 instead of 4317.
      - name: OTEL_EXPORTER_OTLP_ENDPOINT
        value: https://simplest-collector.tracing-system.svc.cluster.local:4318
EOF
```

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: apis/v1alpha1/instrumentation_types.go [hunk 1]
```diff
@@ -97,8 +97,36 @@ type Resource struct {
 // Exporter defines OTLP exporter configuration.
 type Exporter struct {
 	// Endpoint is address of the collector with OTLP endpoint.
+	// The TLS is enabled
```
[the remark this answers is not in the document]
Comment (jaronoff97) [reply apis/v1alpha1/instrumentation_types.go:100]: or should we say if the endpoint contains `https` a `tls` block must be specified?

[the remark this answers is not in the document]
Comment (jaronoff97) [reply apis/v1alpha1/instrumentation_types.go:100]: certainly, in the future it would be great if we could just specify a collector name and namespace instead of an arbitrary string, do some validation and then create the certs for both.

File: apis/v1alpha1/instrumentation_types.go [hunk 2]
```diff
@@ -97,8 +97,36 @@ type Resource struct {
...
+	// TLS defines certificates for TLS.
+	// TLS needs to be enabled by specifying https:// scheme in the Endpoint.
+	TLS *TLS `json:"tls,omitempty"`
+}
+
+// TLS defines TLS configuration for exporter.
+type TLS struct {
+	// SecretName defines secret name that will be used to configure TLS on the exporter.
+	// It is user responsibility to create the secret in the namespace of the workload.
+	// The secret should contain keys ca.crt, tls.key, tls.crt
```
Comment (jaronoff97) [inline apis/v1alpha1/instrumentation_types.go:113]: should or must?

File: apis/v1alpha1/instrumentation_types.go [hunk 3]
```diff
@@ -97,8 +97,36 @@ type Resource struct {
...
+	// CA defines the key of certificate in the configmap map, secret or absolute path to a certificate.
+	// The absolute path can be used when certificate is already present on the workload filesystem e.g.
+	// /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
+	CA string `json:"ca,omitempty"`
+	// Cert defines the key of the client certificate in the secret or absolute path to a certificate.
+	// The absolute path can be used when certificate is already present on the workload filesystem.
+	Cert string `json:"cert,omitempty"`
+	// Key defines a key of the private key in the secret or absolute path to a certificate.
+	// The absolute path can be used when certificate is already present on the workload filesystem.
+	Key string `json:"key,omitempty"`
```
Comment (jaronoff97) [inline apis/v1alpha1/instrumentation_types.go:129]: should we hint the names of the files these should be?

File: pkg/instrumentation/exporter.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	if exporter.TLS == nil {
+		return
+	}
+	// the name cannot be longer than 63 characters
+	secretVolumeName := naming.Truncate("%s-secret-%s", 63, volumeName, exporter.TLS.SecretName)
+	secretMountPath := fmt.Sprintf("/otel-auto-instrumentation-secret-%s", exporter.TLS.SecretName)
+	configMapVolumeName := naming.Truncate("%s-configmap-%s", 63, volumeName, exporter.TLS.ConfigMapName)
+	configMapMountPath := fmt.Sprintf("/otel-auto-instrumentation-configmap-%s", exporter.TLS.ConfigMapName)
+
+	if exporter.TLS.CA != "" {
```
Comment (jaronoff97) [inline pkg/instrumentation/exporter.go:46]: should we require that if one of these is set, all of them should be set?

File: apis/v1alpha1/instrumentation_types.go (hunk 2 above)
Comment (frzifus) [reply apis/v1alpha1/instrumentation_types.go:113]: I think should. Since the CA is not always required, right?

File: apis/v1alpha1/instrumentation_types.go [hunk 5]
```diff
@@ -97,8 +97,36 @@ type Resource struct {
...
+	// The secret should contain keys ca.crt, tls.key, tls.crt
+	SecretName string `json:"secretName,omitempty"`
+
+	// ConfigMapName defines configmap name with CA certificate. If it is not defined CA certificate will be
+	// used from the secret defined in SecretName.
+	ConfigMapName string `json:"configMapName,omitempty"`
+
+	// CA defines the key of certificate in the configmap map, secret or absolute path to a certificate.
+	// The absolute path can be used when certificate is already present on the workload filesystem e.g.
+	// /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
```
Comment (frzifus) [inline apis/v1alpha1/instrumentation_types.go:122]:
```suggestion
	// /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
	// +optional

File: apis/v1alpha1/instrumentation_types.go [hunk 6]
```diff
@@ -97,8 +97,36 @@ type Resource struct {
...
+
+// TLS defines TLS configuration for exporter.
+type TLS struct {
+	// SecretName defines secret name that will be used to configure TLS on the exporter.
+	// It is user responsibility to create the secret in the namespace of the workload.
+	// The secret should contain keys ca.crt, tls.key, tls.crt
+	SecretName string `json:"secretName,omitempty"`
+
+	// ConfigMapName defines configmap name with CA certificate. If it is not defined CA certificate will be
+	// used from the secret defined in SecretName.
```
Comment (frzifus) [inline apis/v1alpha1/instrumentation_types.go:117]:
```suggestion
	// used from the secret defined in SecretName.
```

File: pkg/instrumentation/podmutator.go [hunk 7]
```diff
@@ -460,3 +465,76 @@ func (pm *instPodMutator) selectInstrumentationInstanceFromNamespace(ctx context
...
+		}
+	}
+
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+	return nil
+}
+
+func (pm *instPodMutator) validateInstrumentation(ctx context.Context, inst v1alpha1.Instrumentation, podNamespace string) error {
```
Comment (frzifus) [inline pkg/instrumentation/podmutator.go:518]:
```suggestion
func (i *instPodMutator) validateInstrumentation(ctx context.Context, inst v1alpha1.Instrumentation, podNamespace string) error {
```

File: apis/v1alpha1/instrumentation_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (frzifus) [reply apis/v1alpha1/instrumentation_types.go:100]: >  certainly, in the future it would be great if we could just specify a collector name and namespace instead of an arbitrary string, do some validation and then create the certs for both. 

We should probably offer both. Otherwise it would no longer be possible to report to a custom service or a self-deployed collector?

File: apis/v1alpha1/instrumentation_types.go (hunk 2 above)
Comment (pavolloffay) [reply apis/v1alpha1/instrumentation_types.go:113]: correct, I will rephrase it to make it clear

File: pkg/instrumentation/podmutator.go (hunk 7 above)
Comment (pavolloffay) [reply pkg/instrumentation/podmutator.go:518]: pm is used across the file

File: apis/v1alpha1/instrumentation_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pavolloffay) [reply apis/v1alpha1/instrumentation_types.go:100]: more direct integration with collector will be also needed if we want to support auto-provisioning of certs.

File: pkg/instrumentation/exporter.go (hunk 4 above)
Comment (pavolloffay) [reply pkg/instrumentation/exporter.go:46]: There are three valid combinations:

1. ca
2. key, secret
3. ca, key, secret

The rule here is - if the key is defined the secret must be as well. However I would not implement validation on this to give users more flexibility. Eg. cert/key already defined in the workload or consumed from env var.

File: apis/v1alpha1/instrumentation_types.go (hunk 5 above)
Comment (pavolloffay) [reply apis/v1alpha1/instrumentation_types.go:123]: @frzifus what does the `+optional` do ? It does not seem to have any effect on the generated bundle or code.

We should get rid of it https://book.kubebuilder.io/reference/markers I will submit a follow up PR to remove `+optional`

File: apis/v1alpha1/instrumentation_types.go [hunk 8]
```diff
@@ -97,8 +97,37 @@ type Resource struct {
...
+type TLS struct {
+	// SecretName defines secret name that will be used to configure TLS on the exporter.
+	// It is user responsibility to create the secret in the namespace of the workload.
+	// The secret must contain client certificate (Cert) and private key (Key).
+	// The CA certificate might be defined in the secret or in the config map.
+	SecretName string `json:"secretName,omitempty"`
+
+	// ConfigMapName defines configmap name with CA certificate. If it is not defined CA certificate will be
+	// used from the secret defined in SecretName.
+	ConfigMapName string `json:"configMapName,omitempty"`
```
Comment (iblancasa) [inline apis/v1alpha1/instrumentation_types.go:119]: I'm not sure about this name because it is something associated to the CA directly. Maybe `configMapName` -> `caConfigMap`.

Comment (iblancasa) [inline apis/v1alpha1/instrumentation_types.go:119]: I'm not sure about this name because it is something associated to the CA directly. Maybe `configMapName` -> `caConfigMap`.

File: pkg/instrumentation/exporter.go (hunk 4 above)
Comment (iblancasa) [reply pkg/instrumentation/exporter.go:46]: >  However I would not implement validation on this to give users more flexibility. Eg. cert/key already defined in the workload or consumed from env var.

I'm fine with that but I would show a warning in the webhook validator.

File: pkg/instrumentation/podmutator.go [hunk 9]
```diff
@@ -460,3 +465,76 @@ func (pm *instPodMutator) selectInstrumentationInstanceFromNamespace(ctx context
...
+	if inst.Nginx.Instrumentation != nil {
+		if err := pm.validateInstrumentation(ctx, *inst.Nginx.Instrumentation, podNamespace); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if inst.Sdk.Instrumentation != nil {
+		if err := pm.validateInstrumentation(ctx, *inst.Sdk.Instrumentation, podNamespace); err != nil {
+			errs = append(errs, err)
+		}
+	}
```
Comment (iblancasa) [inline pkg/instrumentation/podmutator.go:510]:
```suggestion
	instrumentations := []struct {
		instrumentation *Instrumentation
	}{
		{inst.Java.Instrumentation},
		{inst.Python.Instrumentation},
		{inst.NodeJS.Instrumentation},
		{inst.DotNet.Instrumentation},
		{inst.Go.Instrumentation},
		{inst.ApacheHttpd.Instrumentation},
		{inst.Nginx.Instrumentation},
		{inst.Sdk.Instrumentation},
	}

	for _, i := range instrumentations {
		if i.instrumentation != nil {
			if err := pm.validateInstrumentation(ctx, *i.instrumentation, podNamespace); err != nil {
				errs = append(errs, err)
			}
		}
	}
```

File: pkg/instrumentation/exporter.go (hunk 4 above)
Comment (pavolloffay) [reply pkg/instrumentation/exporter.go:46]: added

File: apis/v1alpha1/instrumentation_webhook.go [hunk 10]
```diff
@@ -236,6 +236,14 @@ func (w InstrumentationWebhook) validate(r *Instrumentation) (admission.Warnings
 	default:
 		return warnings, fmt.Errorf("spec.sampler.type is not valid: %s", r.Spec.Sampler.Type)
 	}
+
+	if r.Spec.Exporter.TLS != nil {
```
Comment (jaronoff97) [inline apis/v1alpha1/instrumentation_webhook.go:240]: let's also add a validation that if the scheme for the endpoint contains `https` TLS must be defined.

File: pkg/instrumentation/exporter_test.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+func TestExporter(t *testing.T) {
+	tests := []struct {
+		name     string
+		exporter v1alpha1.Exporter
+		expected corev1.Pod
+	}{
+		{
+			name: "ca, crt and key from secret",
+			exporter: v1alpha1.Exporter{
+				Endpoint: "http://collector:4318",
```
Comment (jaronoff97) [inline pkg/instrumentation/exporter_test.go:35]:
```suggestion
				Endpoint: "https://collector:4318",
```

File: pkg/instrumentation/exporter.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+				Value: exporter.Endpoint,
+			})
+		}
+	}
+	if exporter.TLS == nil {
+		return
+	}
+	// the name cannot be longer than 63 characters
+	secretVolumeName := naming.Truncate("otel-auto-secret-%s", 63, exporter.TLS.SecretName)
+	secretMountPath := fmt.Sprintf("/otel-auto-instrumentation-secret-%s", exporter.TLS.SecretName)
```
Comment (jaronoff97) [inline pkg/instrumentation/exporter.go:42]: should we care about the max for a mount path? I'm not 100% there is one (i couldn't find anything in the kube API docs) but given we truncate the names here, i feel like we should try to keep the mount path name minimal too.

File: pkg/instrumentation/exporter_test.go [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+								},
+							},
+							Env: []corev1.EnvVar{
+								{
+									Name:  "OTEL_EXPORTER_OTLP_ENDPOINT",
+									Value: "http://collector:4318",
+								},
+								{
+									Name:  "OTEL_EXPORTER_OTLP_CERTIFICATE",
+									Value: "/otel-auto-instrumentation-configmap-ca-bundle/ca.crt",
```
Comment (jaronoff97) [inline pkg/instrumentation/exporter_test.go:142]: Can someone not keep the cert in a secret?

File: tests/e2e-instrumentation/instrumentation-java-tls/chainsaw-test.yaml [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        file: 00-install-instrumentation.yaml
+  - name: step-01
+    try:
+    - apply:
+        file: 01-install-app.yaml
+    - assert:
+        file: 01-assert.yaml
+    catch:
+      - podLogs:
+          selector: app=my-java
```
Comment (jaronoff97) [inline tests/e2e-instrumentation/instrumentation-java-tls/chainsaw-test.yaml:46]: should we run a curl against the collector to see that span traffic made it successfully to the collector?

File: apis/v1alpha1/instrumentation_webhook.go (hunk 10 above)
Comment (pavolloffay) [reply apis/v1alpha1/instrumentation_webhook.go:240]: no, this can break already existing installations. Users could configure env vars in the `env` section or directly on their deployments.

File: pkg/instrumentation/exporter_test.go (hunk 13 above)
Comment (pavolloffay) [reply pkg/instrumentation/exporter_test.go:142]: certs should be in secret this is how cert-manager and OCP works. 

ca-bundle can be in both.

File: pkg/instrumentation/exporter.go (hunk 12 above)
Comment (pavolloffay) [reply pkg/instrumentation/exporter.go:42]: that does not seem to be an issue only the volume name

File: tests/e2e-instrumentation/instrumentation-java-tls/chainsaw-test.yaml (hunk 14 above)
Comment (pavolloffay) [reply tests/e2e-instrumentation/instrumentation-java-tls/chainsaw-test.yaml:46]: We don't do this for auto-instrumentation tests there is a booked ticket for it  https://github.com/open-telemetry/opentelemetry-operator/issues/552

Comment (jaronoff97) [reply tests/e2e-instrumentation/instrumentation-java-tls/chainsaw-test.yaml:46]: okay fine :P I can do this post my refactor.

File: apis/v1alpha1/instrumentation_webhook.go (hunk 10 above)
Comment (jaronoff97) [reply apis/v1alpha1/instrumentation_webhook.go:240]: let's go the reverse and if TLS is set we should block when https is not set. If HTTPS is set, TLS doesn't need to be set.

**Review by jaronoff97 — APPROVED** [review]

**Review by frzifus — APPROVED** [review]
